docs: adding incident-5 (#2)
decentraland/rca: Incidents/incident-5.md

# Incident 5

|                            |             |
| -------------------------: | :---------- |
| **Reported on**            | July-12-2022 |
| **Mitigation**             | July-12-2022 |
| **Solution Completed**     | July-15-2022 |
| **Severity**               | SEV-3        |

## What happened? 

After updating **peer-ec1** and **peer-ap1** from Catalyst `v4.8.0` to `v4.8.4` the nodes started to throw CPU alarms and the Lambdas Service entered in a continuous reboot cycle. 
 
## Why did it happen?

After several tests, it was detected that the change causing this issue was a [refactor](https://github.com/decentraland/catalyst/pull/1168) on the Lambdas routes to use Well Known Components but the rollout converged with a major DB change that led to think the issue was there and this situation added complexity to the troubleshooting process. 

## Issue Timeline 

Several changes were accumulated in the release rollout and those potential sources of the issues added complexity to finding out what was the root cause of the problem.  

> July-12-2022 - New Catalyst version [v4.8.4](https://github.com/decentraland/catalyst/releases/tag/4.8.4) created. This version had a major change in the DB schema and all the queries impacted by this.

Reviewer: ```suggestion
July-12-2022 - New Catalyst version [v4.8.4](https://github.com/decentraland/catalyst/releases/tag/4.8.4) created. This version had a major change in the DB schema and all the queries impacted by this.

```

> July-12-2022 - Upgrade nodes **peer-ec1** and **peer-ap1** from `v4.8.0` to `v4.8.4`

Reviewer: ```suggestion
July-12-2022 - Upgrade nodes **peer-ec1** and **peer-ap1** from `v4.8.0` to `v4.8.4`

```

> July-12-2022 - After a while of the version update, alarms started to be triggered and lambdas services started to reboot

Reviewer: ```suggestion
July-12-2022 - After a while of the version update, alarms started to be triggered and lambdas services started to reboot

```

> July-12-2022 - To mitigate the issue **peer-ec1** was rollback to `v4.8.0` and the `maxUsers` of **peer-ap1** was set to ``0`` to avoid users being sent to the node

Reviewer: ```suggestion
July-12-2022 - To mitigate the issue **peer-ec1** was rolled back to `v4.8.0` and the `maxUsers` of **peer-ap1** was set to ``0`` to avoid users being sent to the node

```

*CloudFlare Load Balancer*:
![LB](img/incident-5/lb.png)

July-12-2022 The following patterns were detected: 
First the `get_historical_deployments` query started to take a lot of time, leading to think that the changes on the DB had something to do with this issue 

![historic](img/incident-5/historical-deployments.png)

On the other hand, this was also causing CPU spikes 

![cpu](img/incident-5/CPU-Spikes.png)

An the following HTTP Requests pattern started to appear on the Content Server  

![http](img/incident-5/Requests-count.png)

As a result, the AWS instance run out of CPU credits and this didn't allow the node to recover from this situation. 

![CPU](img/incident-5/CPU.png)

And the queries started to throw timeout errors on the lambdas logs: 
``` 
Error: Query read timeout
    at Timeout._onTimeout (/app/node_modules/pg/lib/client.js:526:21)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
```

> July-13-2022 A potential [fix](https://github.com/decentraland/catalyst/commit/3f12deb2536bf6dab39ae7bcdfdfcacb1e71f803) for the DB changes was added and tested on **peer-ap1** but the same issue happened. 

Reviewer: ```suggestion
July-13-2022 A [fix](https://github.com/decentraland/catalyst/commit/3f12deb2536bf6dab39ae7bcdfdfcacb1e71f803) for the DB changes was added and tested on **peer-ap1** but the same issue happened. 
```

> July-13-2022 A Catalyst tag version with a rollback of all the Catalyst DB changes was created and tested on **peer-ap1**. The HTTP Request pattern still happen. Concluding that the issue was not being caused by the DB changes. 

Reviewer: ```suggestion
July-13-2022 A Catalyst tag version with a rollback of all the Catalyst DB changes was created and tested on **peer-ap1**. The HTTP Request pattern was still happening, concluding that the issue was not being caused by the DB changes. 
```

July-13-2022 In order to be sure, another tag version was tested started from `v4.8.0` and with the DB changes and everything worked fine

> July-13-2022 The remaining biggest change was in lambdas, this was rollback and all the changes in the middle patches between `v4.8.0` and `v4.8.4` versions were added, creating a new version with all the changes but this refactor. In theory, the removed change shouldn't have affected the performance as it was just a reorganization of the endpoints to use Well Known Components and not an actual implementation.

Reviewer: ```suggestion
July-13-2022 The remaining biggest change was in lambdas, this was rolled back and all the changes in the middle patches between `v4.8.0` and `v4.8.4` versions were added, creating a new version with all the changes but this refactor. In theory, the removed change shouldn't have affected the performance as it was just a reorganization of the endpoints in preparation for a further refactor to use Well Known Components and not an actual implementation change.
```

                                "g": 0.22265625,
                                "b": 0.1953125,
                                "a": 1
                            }
                        },
                        "hair": {
                            "color": {
                                "r": 0.55078125,
                                "g": 0.125,
                                "b": 0.078125,
                                "a": 1
                            }
                        },
                        "skin": {
                            "color": {
                                "r": 0.94921875,
                                "g": 0.76171875,
                                "b": 0.6484375,
                                "a": 1
                            }
                        }
                    },
                    "tutorialStep": 256,
                    "interests": [],
                    "unclaimedName": "iceloops",
                    "hasConnectedWeb3": true
                }
            ]
        }
    }
]
```

July-15-2022 New [hotfix](https://github.com/decentraland/catalyst/releases/tag/4.8.6) version created and updated.

## Impact 

> No major impact. The Catalyst **peer-ec1** was quickly rollback to `v4.8.0` and **peer-ap1** node was left out from the Load Balancer and the Realm scoring logic to avoid traffic being request/sent from/to that node. 

Reviewer: ```suggestion
No major impact. The Catalyst **peer-ec1** was quickly rolled back to `v4.8.0` and **peer-ap1** node was left out from the Load Balancer and the Realm scoring logic to avoid traffic being request/sent from/to that node. 
```

## Solution 

- Lambdas refactor rollback. 
- Analyzing changing the AWS instance Type to be less restricted by the CPU Credits, there are newer, better, cheaper instance types.

> - Health check configuration fixed on the Load Balancer

Reviewer: ```suggestion
- Health check configuration fixed on the Load Balancer.
```

> - Under research but not part of this issue, the set of accumulated Failed Deployments on **peer-ap1** due to the missing thumbnail 

Reviewer: ```suggestion
- Under research but not part of this issue, the set of accumulated Failed Deployments on **peer-ap1** due to the missing thumbnail.
```

> - Rollout all patches version as soon as they are ready

Reviewer: ```suggestion
- Rollout all patches version as soon as they are ready.
```
